test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java

```diff
@@ -124,8 +134,58 @@
     private static DeepClone deepClone;
     //needed for @Nested
     private static Deque<Class<?>> currentTestClassStack = new ArrayDeque<>();
+    private static ScheduledExecutorService hangDetectionExecutor;
+    private static Duration hangTimeout;
+    private static ScheduledFuture<?> hangTaskKey;
+    private static final Runnable hangDetectionTask = new Runnable() {
+
+        final AtomicBoolean runOnce = new AtomicBoolean();
+
+        @Override
+        public void run() {
+            if (!runOnce.compareAndSet(false, true)) {
+                return;
+            }
+            System.err.println("@QuarkusTest has detected a hang, as there has been no test activity in " + hangTimeout);
+            System.err.println("To configure this timeout use the " + QUARKUS_TEST_HANG_DETECTION_TIMEOUT + " config property");
+            System.err.println("A stack track is below to help diagnose the potential hang");
+            System.err.println("=== Stack Trace ===");
+            ThreadInfo[] threads = ManagementFactory.getThreadMXBean().dumpAllThreads(true, true);
+            for (ThreadInfo info : threads) {
+                if (info == null) {
+                    System.err.println("  Inactive");
+                    continue;
+                }
+                Thread.State state = info.getThreadState();
+                System.err.println("Thread " + info.getThreadName() + ": " + state);
+                if (state == Thread.State.WAITING) {
+                    System.err.println("  Waiting on " + info.getLockName());
+                } else if (state == Thread.State.BLOCKED) {
+                    System.err.println("  Blocked on " + info.getLockName());
+                    System.err.println("  Blocked by " + info.getLockOwnerName());
+                }
+                System.err.println("  Stack:");
+                for (StackTraceElement frame : info.getStackTrace()) {
+                    System.err.println("    " + frame.toString());
+                }
+            }
+            System.err.println("=== End Stack Trace ===");
+            //we only every dump once
+        }
+    };
 
     private ExtensionState doJavaStart(ExtensionContext context, Class<? extends QuarkusTestProfile> profile) throws Throwable {
+        hangDetectionExecutor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
Maybe we could skip the executor completely if  `-Dquarkus.test.hang-detection-timeout=-1`?

Reply:
That make sense, but in what scenario do you envision doing that?

Reply:
Just in case, if there are any problems with the executor/cancelling the task or something like that...

Reply:
I guess it can't hurt :)